nodestream/pipeline/pipeline.py

```diff
@@ -78,9 +140,50 @@ async def work_body(self):
                 self.progress_reporter.report(index, record)
 
     async def work_loop(self):
-        self.start()
-        await self.work_body()
-        await self.stop()
+        # Start only calls some callbacks on the reporter.
+        # If this fails, we can just log the error and move on.
+        try:
+            self.start()
+        except Exception as start_exception:
+            self.logger.exception(
+                f"Exception during start for step {self.step.__class__.__name__}. This will not be considered fatal.",
+                stack_info=True,
+                extra={"step": self.step.__class__.__name__},
+            )
+            self.exceptions["Exception in Start Process:"] = start_exception
+
+        # During the main exection of the code, we want to catch any
+        # exceptions that happen and log them.
+        # Since this step can no longer do any work, we will raise the
+        # exception so that the pipeline can safely come to a stop.
+        try:
+            await self.work_body()
+        except Exception as work_body_exception:
+            self.logger.exception(
+                f"Exception during the work body for step {self.step.__class__.__name__}. This will not be considered fatal.",
+                stack_info=True,
+                extra={"step": self.step.__class__.__name__},
+            )
+            self.exceptions["Exception in Work Body:"] = work_body_exception
+
+        # When we're done, we need to try to call stop on the step.
+        # This is because the step may have some cleanup to do.
+        # If this fails, we are near the end of the exection of the pipeline
+        # so we can just log the error and move on.
+        finally:
+            try:
+                await self.stop()
+            except Exception as stop_exception:
+                self.logger.exception(
+                    f"Exception during stop for step {self.step.__class__.__name__}. This will not be considered fatal.",
+                    stack_info=True,
+                    extra={"step": self.step.__class__.__name__},
+                )
+                self.exceptions["Exception in Stop Process:"] = stop_exception
```

Comment:
I think its better to make these values such as `"Exception in Stop Process:"` Constants defined at the top of the file.